src/qforte/gate.cc

```diff
@@ -27,25 +27,67 @@ size_t Gate::control() const { return control_; }
         throw std::runtime_error("Target index is too large for specified nqbits!");
     }
+    if (control_ >= nqubit) {
+        throw std::runtime_error("Control index is too large for specified nqbits!");
+    }
 
+    size_t nbasis = std::pow(2, nqubit);
     SparseMatrix Spmat = SparseMatrix();
 
-    for (size_t i = 0; i < 2; i++) {
-        for (size_t j = 0; j < 2; j++) {
-            auto op_i_j = gate_[i][j];
-            if (std::abs(op_i_j) > 1.0e-16) {
-                for (size_t I = 0; I < nbasis; I++) {
-                    QubitBasis basis_I = QubitBasis(I);
-                    if (basis_I.get_bit(target_) == j) {
-                        QubitBasis basis_J = basis_I;
-                        basis_J.set_bit(target_, i);
-                        Spmat.set_element(basis_J.index(), basis_I.index(), op_i_j);
+    if (target_ == control_) {
+        // single-qubit case
+        // Iterate over the gate matrix elements
+        for (size_t i = 0; i < 2; i++) {
+            for (size_t j = 0; j < 2; j++) {
+                auto op_i_j = gate_[i][j];
+                // Consider only non-zero elements (within a tolerance)
+                if (std::abs(op_i_j) > 1.0e-14) {
+                    // Iterate over all basis states
+                    for (size_t I = 0; I < nbasis; I++) {
+                        QubitBasis basis_I = QubitBasis(I);
+                        // Check if the target qubit is in the correct state
+                        if (basis_I.get_bit(target_) == j) {
+                            QubitBasis basis_J = basis_I;
+                            // Set the target qubit to the new state
+                            basis_J.set_bit(target_, i);
+                            // Set the corresponding element in the sparse matrix
+                            Spmat.set_element(basis_J.index(), basis_I.index(), op_i_j);
+                        }
+                    }
+                }
+            }
+        }
+    } else {
+        // two-qubit gate
+        for (size_t i = 0; i < 4; i++) {
+            for (size_t j = 0; j < 4; j++) {
+                auto op_i_j = gate_[i][j];
+                if (std::abs(op_i_j) > 1.0e-14) {
+                    for (size_t I = 0; I < nbasis; I++) {
+                        QubitBasis basis_I = QubitBasis(I);
+                        size_t target_bit = basis_I.get_bit(target_);
+                        size_t control_bit = basis_I.get_bit(control_);
+
+                        // Combine the target and control bits into a 2-bit state
+                        size_t combined_state = (control_bit << 1) | target_bit;
+
+                        // Check if the combined state matches the gate matrix column index
+                        if (combined_state == j) {
+                            QubitBasis basis_J = basis_I;
+
+                            // Extract the new states for control and target bits
+                            size_t new_control_bit = (i >> 1) & 1;
```

Comment:
How can the control qubit change? Because it's the control qubit, shouldn't we be able to assume that matrix elements that switch the control qubit are all zero?

I'm _assuming_ that this function is designed to handle any two-qubit gate, and the language of control/target is just a convenience for the most common case.

Reply:
All of your statements are correct. Currently, all but two two-qubit gates implemented in QForte are controlled. The exceptions are the number-preserving `A` gate and the `SWAP` gate. The target/control designation is used to reflect the most common case.